### PR TITLE
CLEWS-21846 Catch get_data connection errors and retry

### PR DESCRIPTION
**Pull request by itsjohnward**
* Related BatchClient improvement: https://github.com/gro-intelligence/api-client/pull/177
* Previous `get_data` retries improvement for API response codes (not exceptions): https://github.com/gro-intelligence/api-client/pull/175

Now get_data catches exceptions, including but not limited to:

* `requests.exceptions.ProxyError`
* `requests.exceptions.Timeout`

And it will treat those similarly to how it treats non-200 status code API responses: automatically retry, and if number of retries exceeds the max, raise an APIError.

### CODE REVIEW

File: api/client/lib.py [hunk 1]
```diff
@@ -147,29 +147,33 @@ def get_data(url, headers, params=None, logger=None):
...
-            params = redirect(params, data.json()['data'][0])
-        elif data.status_code in [404, 401, 500]:
-            break
-        else:
-            logger.error(data.text, extra=log_record)
-    raise Exception('Giving up on {} after {} tries. Error is: {}.'.format(
-        url, retry_count, data.text))
+    raise Exception(
+        'Giving up on {} after {} tries. Error is: {}.'.format(url, log_record['retry_count'], data.text)
+    )
```
Comment (nemozen) [inline api/client/lib.py]: let's keep the style consistent wrt parentheses and line break i.e. closing bracket should not get its own line.

**Review by nemozen — APPROVED** [review]

**Comment by itsjohnward** [comment]
Changed base branch to https://github.com/gro-intelligence/api-client/pull/175 so as to not conflict with changes there

Some extra commits showing here I think should go away when that one's merged and this one's base is switched back to development

**Comment by itsjohnward** [comment]
Merged in the https://github.com/gro-intelligence/api-client/pull/175 changes and now this PR's significantly different from when it was originally approved, so re-requesting review

**Review by muzigao — APPROVED** [review]
looks good to me

**Review by nemozen — APPROVED** [review]

**Pull request merged**